### ISSUE TITLE
Knex is not clearing timeouts after they have become unnecessary

### ISSUE THREAD

**Issue by prk3** [knex/knex#3699]
# Environment

Knex version: 0.20.10
Database + version: postgres 12.1
OS: Linux
Node: 12.15.0
Jest: 25.1.0

# Bug

Jest prints a warning about open handles when a test creates and rollbacks a transaction. When I comment out transacting code, the warning disappears. The connection to postgres database is closed in afterAll call.

```
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your
 tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

`--detectOpenHandles` flag does not detect anything.

Here is the reproduction code:
```javascript
const Knex = require('knex');

const db = Knex({
  client: process.env.DB_TYPE || 'pg',
  connection: {
    host: process.env.DB_HOST || 'localhost',
    port: Number(process.env.DB_PORT) || 5432,
    user: process.env.DB_USER || 'postgres',
    password: process.env.DB_USER || 'postgres',
    database: process.env.DB_DATABASE || 'postgres',
  },
});

describe('test', () => {
  afterAll(async () => {
    await db.destroy();
  });

  it('transacts', async () => {
    const t = await db.transaction();

    await t.rollback();
  });
});
```

**Comment by elhigu** [comment]
Is it only with transactions or does making any query provoke this? Does it happen also if you commit transaction correctly?

Also that transaction code is  invalid. You should use knex.transactionProvider for that kind of scenario.

```
  it('transacts', async () => {   
    const trxProvider = knex.transactionProvider();
    const trx = await trxProvider();
    trx.rollback(new Error('Fail'));
  
    try {
      await trx.executionPromise;
    } catch (err) {
      // just to prevent unhandled rejection caused by rollback
    }
  }); 
```

or just the safer handler way with implicit commit / rollback , which works for most use cases just fine:

 ```
  it('transacts', async () => {
     try {
        // just implicitly rolls back because of thrown exception
        await db.transaction(async trx => {
           threw now Error("will implicitly rollback because handler did reject");
        }); 
     } catch (err) {
        // to prevent unhandled rejection
     }
  }); 
```

**Comment by elhigu** [comment]
Sorry I just read the issue again... I'm pretty sure this just happens because of invalid use of knex. Might be that await db.destroy() doesn't complete because of invalid use of transactions there might be connection leased from pool that is never returned there. 

Please try to fix your transaction code and if problem still  persists lets reopen this.

**Comment by prk3** [comment]
I am pretty sure the code is in par with knex documentation (transaction objects). Anyway, I fixed the code and did a bit more testing. Queries and committing works fine, but rolling back makes jest throw `Jest did not exit one second after the test run has completed.`. 

```javascript
it('transacts', async () => {
  try {
    await db.transaction(async (trx) => {
      await trx('recipes').select('*');
      expect('a').toHaveLength(1);
      // commenting out the line below fixes the test
      throw new Error('will implicitly rollback because handler did reject');
    });
  } catch (err) {
    // do nothing
  }
});
```

This code gives exactly the same result:
```javascript
it('transacts', async () => {
  const trxProvider = db.transactionProvider();
  const trx = await trxProvider();
  await trx('recipes').select('*');
  expect('a').toHaveLength(1);
  trx.rollback(new Error('Fail'));

  try {
    await trx.executionPromise;
  } catch (err) {
    console.log('Expected rollback');
  }
});
```

So it seems that rollback leaves an open handle.

**Comment by elhigu** [comment]
> I am pretty sure the code is in par with knex documentation (transaction objects).

huh, there is actually usage like that... need to verify if that really works as intended. I would be expecting at least some uncaught promise rejection errors... 

If you run that code without jest does the code exit? If so, how long it takes to exit?

**Comment by briandamaged** [comment]
@prk3 : Also, you might need to provide the `doNotRejectOnRollback` parameter when creating your transaction.  Ex:

```
const t = await db.transaction(undefined, {doNotRejectOnRollback: false});
```

(It seems that the `doNotRejectOnRollback` parameter was introduced to help maintain backwards-compatibility with an older version of `knex`.)

**Comment by prk3** [comment]
@elhigu Code without jest:
```javascript
const Knex = require('knex');

const db = Knex({
  ...
});

(async () => {
  try {
    await db.transaction(async (trx) => {
      await trx('recipes').select('*');
      throw new Error('rollback!');
    });
  } catch (err) {
    // no nothing
  } finally {
    await db.destroy();
    console.log((new Date).toISOString());
  }
})();
```

Time measurements:
```
> node a.js && date
2020-03-04T19:23:33.581Z
Wed  4 Mar 19:23:38 GMT 2020
>
```

If I comment out throw...

```
> node a.js && date
2020-03-04T19:34:24.423Z
Wed  4 Mar 19:34:24 GMT 2020
>
```

So if a rollback happens, the node process lives for 5 more seconds second after `db` is destroyed.

@briandamaged I run `npm install git+ssh://git@github.com/knex/knex.git#master`. This should install knex from master, correct?

The result is exactly the same: rollback leaves the process running for extra 5 seconds. `doNotRejectOnRollback` had no effect neither: no matter if it was true of false, rollback was triggering the warning, commit was not.

**Comment by briandamaged** [comment]
@prk3 : Alternative hypothesis: the current `knex` code actually creates a 5-second delay whenever a `rollback` is issued.  If it detects that the `rollback` has not completed within this timeframe, it then attempts to "cancel the rollback":

https://github.com/knex/knex/blob/51052c52e896aabd50702e92372e928269a19783/lib/transaction.js#L143-L152

My guess is that we're not actually observing "open handles".  Instead, node is still waiting 5 seconds for this `setTimeout(..)` to conclude.  Once that has completed, it finally allows the process to exit.

(Side-note: I suspect that this behavior predates some of the changes to `knex`'s connection pooling logic.  So, this 5-second delay might no longer even be necessary.  We'll need to investigate this further)

**Comment by prk3** [comment]
@briandamaged It makes sense now!

```javascript
let timeoutId;

const timeoutPromise = new Promise((res, rej) => {
  timeoutId = setTimeout(() => rej(new KnexTimeoutException), 5000);
});

const queryPromise = this.query(conn, 'ROLLBACK', 2, error).then(result => {
  clearTimeout(timeoutId);
  return result;
});

return Promise.race([timeoutPromise, queryPromise]).catch(err => { 
  if (!(err instanceof KnexTimeoutError)) { 
    return Promise.reject(err); 
  } 
  this._rejecter(error); 
});
```

This modification would solve my problem. But what if timeout rejects first? In my code the query handler would be cleaned up by db.destory(), but in general case?

**Comment by briandamaged** [comment]
@prk3 .  Hmm... it might be worth investigating the technique you outlined.

On a related note: we've actually removed the `timeout(..)` wrapper from a few other `knex` methods over the last few weeks.  It turns out that some (all?) of them are no longer necessary thanks to the fact that `knex` uses a better pooling library now.  It might turn out that this specific call to `timeout(..)` is also no longer necessary.

(I did a quick experiment to check this, but unfortunately it's causing a handful of unit tests to fail for seemingly unrelated reasons.  So, it's going to take a bit more investigation before we can conclude that the call to `timeout(..)` can safely be removed in this case)

**Comment by elhigu** [comment]
I'm pretty sure that Tarn timeouts will not help in this case, since the connection is already acquired from the pool. Tarn does not have any max lease timeouts which could kick in for connection that has been too long time out of the pool.

**Comment by briandamaged** [comment]
@elhigu : Makes sense.  In that case, we might just need to cancel the `setTimeout(..)` call like @prk3 suggested.

**Comment by briandamaged** [comment]
k -- I just created a PR related to this:

https://github.com/knex/knex/pull/3703

You can find the corresponding branch here:

https://github.com/briandamaged/knex/tree/fix/clear-unnecessary-timeouts

@prk3 : Can you give this branch a try when you have a moment?  Thanks!

Currently, the tests seem to be passing for me.  But, I might add a few more unit tests just to be on the safe side.

**Comment by prk3** [comment]
@briandamaged Good news. I tested my code against the fix branch and jest exited gracefully. :tada:

**Comment by kibertoad** [comment]
@prk3 Released in 0.20.11
